File: codemod.js
const fs = require('fs');
const path = require('path');

function processFile(filepath) {
    let content = fs.readFileSync(filepath, 'utf-8');
    const originalContent = content;

    // Skip files that don't use currentUser or setCurrentUser
    if (!content.includes('currentUser') && !content.includes('setCurrentUser')) {
        return;
    }

    // Skip store.tsx and this script itself
    if (filepath.includes('store.tsx') || filepath.endsWith('.js')) {
        return;
    }

    console.log(`Modifying ${filepath}`);

    // Add useSession import if needed
    if (!content.includes('useSession')) {
        const importRegex = /^import .*;\n?/gm;
        let lastMatch = null;
        let match;
        while ((match = importRegex.exec(content)) !== null) {
            lastMatch = match;
        }
        
        const importStr = 'import { useSession } from "next-auth/react";\n';
        if (lastMatch) {
            content = content.slice(0, lastMatch.index + lastMatch[0].length) + importStr + content.slice(lastMatch.index + lastMatch[0].length);
        } else {
            content = importStr + content;
        }
    }
            
    // Remove currentUser from useAppStore mapping
    content = content.replace(/currentUser,\s*/g, '');
    content = content.replace(/,\s*currentUser/g, '');
    content = content.replace(/setCurrentUser,\s*/g, '');
    content = content.replace(/,\s*setCurrentUser/g, '');
    
    // Remove empty useAppStore
    content = content.replace(/const\s*\{\s*\}\s*=\s*useAppStore\(\);?\n?/g, '');

    // Inject useSession
    if (content.includes('useAppStore()')) {
        content = content.replace(
            /(const\s+\{.*\}\s*=\s*useAppStore\(\);?)/,
            '$1\n    const { data: session } = useSession();\n    const currentUser = session?.user as any;'
        );
    } else {
        // If there's no useAppStore(), we try to inject at the start of the component
        // Typically it matches `export function ... {` or `export default function ... {`
        content = content.replace(
            /(export\s+(?:default\s+)?function\s+\w+\([^)]*\)\s*\{)/,
            '$1\n    const { data: session } = useSession();\n    const currentUser = session?.user as any;'
        );
    }
        
    if (originalContent !== content) {
        fs.writeFileSync(filepath, content, 'utf-8');
    }
}

function walkDir(dir) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const filepath = path.join(dir, file);
        const stat = fs.statSync(filepath);
        if (stat.isDirectory()) {
            walkDir(filepath);
        } else if (filepath.endsWith('.tsx') || filepath.endsWith('.ts')) {
            processFile(filepath);
        }
    }
}

const srcDir = path.join(__dirname, 'src');
walkDir(srcDir);

File: codemod.py
import os
import re

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    # Skip files that don't use currentUser or setCurrentUser
    if 'currentUser' not in content and 'setCurrentUser' not in content:
        return

    # Skip store.tsx and this script itself
    if 'store.tsx' in filepath or filepath.endswith('.py'):
        return

    print(f"Modifying {filepath}")
    original_content = content

    # Add import if needed
    if 'useSession' not in content:
        # insert after last import
        imports = list(re.finditer(r'^import .*;\n', content, re.MULTILINE))
        if imports:
            last_import = imports[-1]
            content = content[:last_import.end()] + 'import { useSession } from "next-auth/react";\n' + content[last_import.end():]
        else:
            content = 'import { useSession } from "next-auth/react";\n' + content
            
    # Remove currentUser from useAppStore
    content = re.sub(r'currentUser,\s*', '', content)
    content = re.sub(r',\s*currentUser', '', content)
    content = re.sub(r'setCurrentUser,\s*', '', content)
    content = re.sub(r',\s*setCurrentUser', '', content)
    
    # If useAppStore() is empty because we removed the only things, we remove it
    content = re.sub(r'const\s*{\s*}\s*=\s*useAppStore\(\);?\n?', '', content)

    # Inject useSession in components
    # Find components (export function ..., export default function ..., const ... = () => ...)
    # This regex is a simple heuristic -> finding the first block after the destructure or function def
    # Alternatively, just inject `const { data: session } = useSession();\n    const currentUser = session?.user as any;` right after `useAppStore()`
    
    # Let's find useAppStore() calls and insert next to it
    if 'useAppStore()' in content:
        content = re.sub(
            r'(const\s+{[^}]*}\s*=\s*useAppStore\(\);?)',
            r'\1\n    const { data: session } = useSession();\n    const currentUser = session?.user as any;',
            content
        )
    else:
        # If no useAppStore but had currentUser (e.g. from props?), skip or try to find a place
        # Actually most are from useAppStore
        pass
        
    if original_content != content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)

src_dir = os.path.join(os.path.dirname(__file__), 'src')
for root, _, files in os.walk(src_dir):
    for filename in files:
        if filename.endswith('.tsx') or filename.endswith('.ts'):
            filepath = os.path.join(root, filename)
            process_file(filepath)
